File: Java/flipAndInvertImage.java
// Solution: methods to flip and invert array
// Runtime: O(n)

class Solution {
    public int[][] flipAndInvertImage(int[][] A) {

        for (int i = 0; i < A.length; i++) {
            flip(A[i]);
        }

        for (int i = 0; i < A.length; i++) {
            invert(A[i]);
        }

        return A;
    }

    public static int[] flip(int[] A) {
        int startIndex = 0;
        int endIndex = A.length-1;

        while (startIndex < endIndex) {
            int temp = A[startIndex];
            A[startIndex] = A[endIndex];
            A[endIndex] = temp;
            startIndex++;
            endIndex--;
        }
        return A;
    }

    public static int[] invert(int[] arr) {

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == 1) {
                arr[i] = 0;
            }
            else {
                arr[i] = 1;
            }
        }
        return arr;
    }
}

File: Golang/flipAndInvertImage.go
// Solution: method that reverses array and method that inverts it. Very intuitive
// Runtime: O(n)

func flipAndInvertImage(A [][]int) [][]int {

    for i := 0; i < len(A); i++ {
        reverseArray(A[i])
    }

    for i := 0; i < len(A); i++ {
        invertArray(A[i])
    }
    return A
}

func reverseArray(arr []int)[]int {

    low := 0
    high := len(arr)-1

    for low < high {
        arr[low], arr[high] = arr[high], arr[low]
        low++
        high--
    }
    return arr
}

func invertArray(arr []int)[]int {

    for i := 0; i < len(arr); i++ {
        if (arr[i] == 1) {
            arr[i] = 0
        } else {
            arr[i] = 1
        }
    }
    return arr
}
